d2common/d2fileformats/d2ds1/ds1_layers.go

```diff
@@ -0,0 +1,432 @@
+	}
+}
+
+func (l *ds1Layers) Width() int {
+	w, _ := l.Size()
+	return w
+}
+
+func (l *ds1Layers) SetWidth(w int) {
+	l.SetSize(w, l.height)
+}
+
+func (l *ds1Layers) Height() int {
+	_, h := l.Size()
+	return h
+}
+
+func (l *ds1Layers) SetHeight(h int) {
+	l.SetSize(l.width, h)
+}
+
+// generic push func for all layer types
+func (l *ds1Layers) push(t layerGroupType, layer *layer) {
+	l.ensureInit()
+	l.cull()
+
+	var group *layerGroup
+
+	var max int
+
+	switch t {
+	case floorLayerGroup:
+		group = &l.Floors
+		max = maxFloorLayers
+	case wallLayerGroup:
+		group = &l.Walls
+		max = maxWallLayers
+	case orientationLayerGroup:
+		group = &l.Orientations
+		max = maxOrientationLayers
+	case shadowLayerGroup:
+		group = &l.Shadows
+		max = maxShadowLayers
+	case substitutionLayerGroup:
+		group = &l.Substitutions
+		max = maxSubstitutionLayers
+	default:
+		return
+	}
+
+	if len(*group) < max {
+		*group = append(*group, layer)
+	}
+}
+
+// generic pop func for all layer types
+func (l *ds1Layers) pop(t layerGroupType) *layer {
+	l.ensureInit()
+	l.cull()
+
+	var group *layerGroup
+
+	var theLayer *layer
+
+	switch t {
+	case floorLayerGroup:
+		group = &l.Floors
+	case wallLayerGroup:
+		group = &l.Walls
+	case orientationLayerGroup:
+		group = &l.Orientations
+	case shadowLayerGroup:
+		group = &l.Shadows
+	case substitutionLayerGroup:
+		group = &l.Substitutions
+	default:
+		return nil
+	}
+
+	// remove last layer of slice and return it
+	if len(*group) > 0 {
+		lastIdx := len(*group) - 1
+		theLayer = (*group)[lastIdx]
+		*group = (*group)[:lastIdx]
+
+		return theLayer
+	}
+
+	return nil
+}
+
+func (l *ds1Layers) get(t layerGroupType, idx int) *layer {
+	l.ensureInit()
+	l.cull()
+
+	var group *layerGroup
+
+	switch t {
+	case floorLayerGroup:
+		group = &l.Floors
+	case wallLayerGroup:
+		group = &l.Walls
+	case orientationLayerGroup:
+		group = &l.Orientations
+	case shadowLayerGroup:
+		group = &l.Shadows
+	case substitutionLayerGroup:
+		group = &l.Substitutions
+	default:
+		return nil
+	}
+
+	if idx >= len(*group) || idx < 0 {
+		return nil
+	}
+
+	return (*group)[idx]
+}
+
+func (l *ds1Layers) insert(t layerGroupType, idx int, newLayer *layer) {
+	l.ensureInit()
+	l.cull()
+
+	if newLayer == nil {
+		return
+	}
+
+	var group layerGroup
+
+	switch t {
+	case floorLayerGroup:
+		group = l.Floors
+	case wallLayerGroup:
+		group = l.Walls
+	case orientationLayerGroup:
+		group = l.Orientations
+	case shadowLayerGroup:
+		group = l.Shadows
+	case substitutionLayerGroup:
+		group = l.Substitutions
+	default:
+		return
+	}
+
+	if len(group) == 0 {
+		group = append(group, newLayer) // nolint:staticcheck // we possibly use group later
+		return
+	}
+
+	if idx > len(group)-1 {
+		idx = len(group) - 1
+	}
+
+	// example:
+	// suppose
+	//		idx=1
+	//		newLayer=c
+	// 		existing layerGroup is [a, b]
+	group = append(group, group[idx:]...) // [a, b] becomes [a, b, b]
+	group[idx] = newLayer                 // [a, b, b] becomes [a, c, b]
```

Comment:
what happens if:
idx=1
group = [a, b, c]
newLayer = d ?
then in append we have:
[a, b, c] + [b, c] so [a, b, c, b, c]
and after changing index:
[a, b, d, b, c]

but result of this operation should be:
[a, b, d, c]

so IMO this code will not work properly

Reply:
let's do just:
```golang
groups = append(groups[:idx], newLayer, groups[idx:]...)
```

Reply:
fixed by 41c1d8e